src/amms/erc_4626/mod.rs

```diff
@@ -0,0 +1,323 @@
+            self.withdraw_fee = 0;
+        // Assuming 18 decimals, if the delta of 1e20 is half the delta of 2e20, relative fee.
+        // Delta / (amount without fee / 10000) to give us the fee in basis points
+        } else if withdraw_fee_delta_1 * U256::from(2) == withdraw_fee_delta_2 {
+            self.withdraw_fee =
+                (withdraw_fee_delta_1 / (withdraw_no_fee / U256::from(10_000))).to();
+        } else {
+            // If not a relative fee or zero, ignore vault
+            todo!("Handle error")
+        }
+
+        // if above does not error => populate the vault
+        self.vault_token = vault_token;
+        self.vault_token_decimals = vault_token_dec as u8;
+        self.asset_token = asset_token;
+        self.asset_token_decimals = asset_token_dec as u8;
+        self.vault_reserve = vault_reserve;
+        self.asset_reserve = asset_reserve;
+
+        Ok(self)
+    }
+}
+
+// TODO: swap calldata
+impl ERC4626Vault {
+    // Returns a new, unsynced ERC4626 vault
+    pub fn new(address: Address) -> Self {
+        Self {
+            vault_token: address,
+            ..Default::default()
+        }
+    }
+
+    pub fn get_amount_out(&self, amount_in: U256, reserve_in: U256, reserve_out: U256) -> U256 {
+        if amount_in.is_zero() {
+            return U256::ZERO;
+        }
+
+        if self.vault_reserve.is_zero() {
+            return amount_in;
+        }
+
+        let fee = if reserve_in == self.vault_reserve {
+            self.withdraw_fee
+        } else {
+            self.deposit_fee
+        };
+
+        amount_in * reserve_out / reserve_in * U256::from(10000 - fee) / U256_10000
+    }
+
+    // TODO: Right now this will return a uv2 error, fix this
+    pub fn calculate_price_64_x_64(&self, base_token: Address) -> Result<u128, AMMError> {
+        let decimal_shift = self.vault_token_decimals as i8 - self.asset_token_decimals as i8;
+
+        // Normalize reserves by decimal shift
+        let (r_v, r_a) = match decimal_shift.cmp(&0) {
+            Ordering::Less => (
+                self.vault_reserve * U256::from(10u128.pow(decimal_shift.unsigned_abs() as u32)),
+                self.asset_reserve,
+            ),
+            _ => (
+                self.vault_reserve,
+                self.asset_reserve * U256::from(10u128.pow(decimal_shift as u32)),
+            ),
+        };
+
+        // Withdraw
+        if base_token == self.vault_token {
+            if r_v.is_zero() {
+                // Return 1 in Q64
+                Ok(U128_0X10000000000000000)
+            } else {
+                Ok(div_uu(r_a, r_v)?)
+            }
+        // Deposit
+        } else if r_a.is_zero() {
+            // Return 1 in Q64
+            Ok(U128_0X10000000000000000)
+        } else {
+            Ok(div_uu(r_v, r_a)?)
+        }
+    }
+
+    pub async fn get_reserves<T, N, P>(
+        &self,
+        provider: P,
+        block_number: u64,
```

Comment:
It's be nice to make this a `BlockNumberOrTag` that way the caller can pass `BlockNumberOrTag::Latest` instead of fetching the latest block.